lib/caching.js

```diff
@@ -66,68 +68,171 @@ var caching = function(args) {
      * Wraps a function in cache. I.e., the first time the function is run,
      * its results are stored in cache so subsequent calls retrieve from cache
      * instead of calling the function.
+     * You can pass any number of keys as long as the wrapped function returns
+     * an array with the same number of values and in the same order.
      *
      * @function
      * @name wrap
      *
-     * @param {string} key - The cache key to use in cache operations
+     * @param {string} key - The cache key to use in cache operations. Can be one or many.
      * @param {function} work - The function to wrap
      * @param {object} [options] - options passed to `set` function
      * @param {function} cb
      *
      * @example
-     *   var key = 'user_' + userId;
-     *   cache.wrap(key, function(cb) {
-     *       User.get(userId, cb);
-     *   }, function(err, user) {
-     *       console.log(user);
-     *   });
+     * var key = 'user_' + userId;
+     * cache.wrap(key, function(cb) {
+     *     User.get(userId, cb);
+     * }, function(err, user) {
+     *     console.log(user);
+     * });
+     *
+     * // Multiple keys
+     * var key = 'user_' + userId;
+     * var key2 = 'user_' + userId2;
+     * cache.wrap(key, key2, function(cb) {
+     *     User.getMany([userId, userId2], cb);
+     * }, function(err, users) {
+     *     console.log(users[0]);
+     *     console.log(users[1]);
+     * });
      */
-    self.wrap = function(key, work, options, cb) {
-        if (typeof options === 'function') {
-            cb = options;
-            options = {};
+    self.wrap = function() {
+        var args = Array.prototype.slice.apply(arguments);
+        var length = args.length;
+        var work;
+        var options = {};
+        var cb;
+
+        /**
+         * As we can receive an unlimited number of keys
+         * we find the index of the first function which is
+         * the "work" handler to fetch the keys.
+         */
+        for (var i = 0; i < length; i += 1) {
+            if (typeof args[i] === 'function') {
+                if (typeof args[i + 2] === 'function') {
+                    cb = args.pop();
+                } else if (typeof args[i + 1] === 'function') {
+                    cb = args.pop();
+                }
+                if (isObject(args[i + 1])) {
+                    options = args.pop();
+                }
+                work = args.pop();
+                break;
+            }
         }
 
         if (!cb) {
-            return wrapPromise(key, work, options);
+            args.push(work);
+            args.push(options);
+            return wrapPromise.apply(this, args);
+        }
+
+        var isMultiple = args.length > 1;
```

Comment:
If you want, from this line I could refactor (and the same in multi_caching):

```js
if (args.length > 1) {
    return wrapMultiple(args, work, options, cb);
}
var key = args;

// Put back the current "wrap" code
```

Reply:
Yeah, I think I like this refactor.